internal/providers/msentraid/msentraid.go

```diff
@@ -143,63 +143,64 @@ func (p Provider) getGroups(token *oauth2.Token) ([]info.Group, error) {
 
 	// Get the groups (only the groups, not directory roles or administrative units, because that would require
 	// additional permissions) which the user is a member of.
-	graphGroups, err := client.Me().TransitiveMemberOf().GraphGroup().Get(context.Background(), nil)
+	graphGroups, err := getAllUserGroups(client)
 	if err != nil {
 		return nil, fmt.Errorf("failed to get user groups: %v", err)
 	}
 
 	var groups []info.Group
-	for _, obj := range graphGroups.GetValue() {
-		unknown := "Unknown"
-		msGroup, ok := obj.(*msgraphmodels.Group)
-		if !ok {
-			id, oType := obj.GetId(), obj.GetOdataType()
-			if id == nil {
-				id = &unknown
-			}
-			if oType == nil {
-				oType = &unknown
-			}
-			slog.Debug(fmt.Sprintf(
-				"Found non-group object with ID: %q of type: %q in graphsdk response. Ignoring it",
-				*id, *oType,
-			))
-			continue
+	for _, msGroup := range graphGroups {
+		idPtr := msGroup.GetId()
+		if idPtr == nil {
+			slog.Warn(pp.Sprintf("Could not get ID for group: %v", msGroup))
+			return nil, errors.New("could not get group id")
 		}
+		id := *idPtr
 
-		v, err := msGroup.GetBackingStore().Get("displayName")
-		if err != nil {
-			return nil, fmt.Errorf("failed to get displayName from group object: %v", err)
+		groupNamePtr := msGroup.GetDisplayName()
+		if groupNamePtr == nil {
+			slog.Warn(pp.Sprintf("Could not get display name for group object (ID: %s): %v", id, msGroup))
+			return nil, errors.New("could not get group name")
 		}
-		name, ok := v.(*string)
-		if !ok || name == nil {
-			id := msGroup.GetId()
-			if id == nil {
-				id = &unknown
-			}
-			slog.Warn(pp.Sprintf("Could not get displayName from group object (ID: %s) found: %v", *id, *msGroup))
-			return nil, errors.New("could not parse group name")
-		}
-		groupName := strings.ToLower(*name)
+		groupName := strings.ToLower(*groupNamePtr)
 
-		// Local group
+		// Check if the group is a local group, in which case we don't set the UGID (because that's how the user manager
+		// differentiates between local and remote groups).
 		if strings.HasPrefix(groupName, localGroupPrefix) {
 			groupName = strings.TrimPrefix(groupName, localGroupPrefix)
 			groups = append(groups, info.Group{Name: groupName})
 			continue
 		}
 
-		v, err = msGroup.GetBackingStore().Get("id")
+		groups = append(groups, info.Group{Name: groupName, UGID: id})
+	}
+
+	return groups, nil
+}
+
+func getAllUserGroups(client *msgraphsdk.GraphServiceClient) ([]msgraphmodels.Groupable, error) {
+	var groups []msgraphmodels.Groupable
```

Comment:
We should move this declaration closer to its use (i.e. line 192). Can we also avoid the `var` directive and go for `:=` instead?